File: deferred/python/test/deferred-tests.py
import community.deferred
reload(community.deferred)
from community.deferred import defer, cancel_all, cancel
from core.log import log_traceback, logging, LOG_PREFIX
from org.joda.time import DateTime
import time

log = logging.getLogger("{}.TEST.deferred".format(LOG_PREFIX))

log.info("Starting deferred tests")

# Create an Item to test with
log.info("Creating test Item Deferred_Test")
from core.items import add_item
item = "Deferred_Test"
add_item(item, item_type="Switch")

try:
    events.postUpdate(item, "OFF")
    time.sleep(0.1)
    assert items[item] == OFF, "Item didn't initialize to OFF"

    # Schedule based on DT
    t = DateTime.now().plusSeconds(1)
    defer(item, "ON", t, log)
    time.sleep(1.1)
    assert items[item] == ON, "Item didn't go to ON after a second with specific time"

    # Schedule based on duration
    defer(item, "OFF", "1s", log)
    time.sleep(1.1)
    assert items[item] == OFF, "Item didn't go to OFF after a second with duration"

    # Reschedule
    defer(item, "ON", "1s", log)
    time.sleep(0.1)
    assert items[item] == OFF, "Item isn't still OFF after initial schedule"
    defer(item, "ON", "2s", log)
    time.sleep(1)
    assert items[item] == OFF, "Timer didn't get rescheduled!"
    time.sleep(1.1)
    assert items[item] == ON, "Timer didn't reschedule on time!"

    # Cancel
    defer(item, "OFF", "1s", log)
    assert items[item] == ON, "Item isn't still ON after last test"
    cancel(item)
    time.sleep(1.1)
    assert items[item] == ON, "Timer didn't cancel!"

    # Cancel All
    defer(item, "OFF", "1s", log)
    cancel_all()
    time.sleep(1.1)
    assert items[item] == ON, "Timer didn't cancel all"

except AssertionError:
    import traceback
    log.error("Exception: {}".format(traceback.format_exc()))
except TypeError:
    import traceback
    log.error("Exception: {}".format(traceback.format_exc()))
else:
    log.info("Deferred tests passed!")
finally:
    log.info("Deleting test Item")
    from core.items import remove_item
    remove_item(item)

File: deferred/javascript/test/deferred-tests.js
var logger = Java.type("org.slf4j.LoggerFactory").getLogger("org.openhab.model.script.Deferred Tests");

var OPENHAB_CONF = java.lang.System.getenv("OPENHAB_CONF");
load(OPENHAB_CONF + "/automation/lib/javascript/community/deferred.js");

logger.info("Creating Deferred object");
var deferred = new Deferred();

/**
 * TODO: Create the Item for the test and then remove it at the end.
 * In the mean time manually create a test Item and update testItem.
 */
var testItem = "aTestSwitch";

logger.info("Starting Deferred tests");
var cont = true;

var assert = function(expected, msg, count) {
  if(items[testItem] != expected) {
    logger.error("[Test " + count + "]: Expected " + expected + " Found " + items[testItem] + ": " + msg);
    return false;
  }
  logger.debug("Test " + count + " passed");
  return true;
}

var reset = function() {
  var cont = true;
  events.postUpdate(testItem, "UNDEF");
  java.lang.Thread.sleep(100);
  return assert(UNDEF, "Failed to initialize " + testItem, "init");
}

// Schedule
cont = reset();
if(cont) {
  deferred.defer(testItem, "ON", "1s");
  java.lang.Thread.sleep(100);
  cont = assert(UNDEF, "Deferred updated " + testItem  + " too soon!", "1a");
  java.lang.Thread.sleep(1100);
  cont = assert(ON, "Failed to update " + testItem + " as scheduled", "1b");
}

// Reschedule
cont = (cont) ? reset() : cont;
if(cont) {
  deferred.defer(testItem, "OFF", "1s");
  java.lang.Thread.sleep(100);
  cont = assert(UNDEF, "Updated " + testItem  + " too soon!", "2a");
}
if(cont) {
  deferred.defer(testItem, "ON", "2s");
  java.lang.Thread.sleep(1000);
  cont = assert(UNDEF, "Deferred didn't reschedule", "2b");
}
if(cont) {
  java.lang.Thread.sleep(1100);
  cont = assert(OFF, "Didn't update " + testItem + " on reschedule", "2c");
}

// Cancel
cont = (cont) ? reset() : cont;
if(cont) {
  deferred.defer(testItem, "ON", "1s");
  java.lang.Thread.sleep(100);
  deferred.cancel(testItem);
  java.lang.Thread.sleep(1100);
  cont = assert(UNDEF, "Didn't cancel!", "3");
}

// Cancel All
cont = (cont) ? reset() : cont;
if(cont) {
  deferred.defer(testItem, "ON", "1s");
  java.lang.Thread.sleep(100);
  deferred.cancelAll();
  java.lang.Thread.sleep(1100);
  assert(UNDEF, "Cancel all didn't cancel", "4");
}

if(cont) {
  logger.info("All tests passed");
}